separate migrations into its own guide (#9064)
supabase/supabase: apps/reference/docs/guides/database/migrating-between-projects.mdx

---
id: migrating-between-projects
title: 'Migrating Between Projects'
description: Migrate data from one project to another
---

> Migrating projects can be achieved using standard PostgreSQL tooling. This is particularly useful for older projects (e.g. to use a newer Postgres version).

Reviewer: ```suggestion
Use standard PostgreSQL tooling to migrate data from one Supabase project to another. This is useful for older projects (e.g., to use a newer Postgres version).
```

Author: @dannykng I think the initial phrasing was to indicate that no custom tooling is required to do the migration and standard postgres tooling can be used to do it and we will be going over how in this guide.

Reviewer: 👍🏼 thanks for clarifying!

## Before you begin

> - Make sure [Postgres](https://www.postgresql.org/download/) is installed so you can run `psql` and `pg_dump`.
> - Create a new Supabase project.
> - If you enabled Database Webhooks on your old project, enable it on your new project.
> - Store the old project's database URL as `$OLD_DB_URL` and the new project's as `$NEW_DB_URL`.

Reviewer: ```suggestion
- Install [Postgres](https://www.postgresql.org/download/) so you can run `psql` and `pg_dump`.
- Create a new [Supabase project](https://app.supabase.com).
- Enable [Database Webhooks](https://app.supabase.com/project/_/database/hooks) in your new project if you enabled them in your old project.
- Store the old project's database URL as `$OLD_DB_URL` and the new project's as `$NEW_DB_URL`.
```

## Migrate the database

> 1. Run `ALTER ROLE postgres SUPERUSER` in the _old_ project's SQL editor
> 2. Run `pg_dump --clean --if-exists --quote-all-identifiers -h $OLD_DB_URL -U postgres > dump.sql` from your terminal
> 3. Run `ALTER ROLE postgres NOSUPERUSER` in the _old_ project's SQL editor
> 4. Run `ALTER ROLE postgres SUPERUSER` in the _new_ project's SQL editor
> 5. Run `psql -h $NEW_DB_URL -U postgres -f dump.sql` from your terminal
> 6. Run `TRUNCATE storage.objects` in the _new_ project's SQL editor
> 7. Run `ALTER ROLE postgres NOSUPERUSER` in the _new_ project's SQL editor

Reviewer: Can we break this up into old project and new project steps? Thoughts?

```suggestion
In your old project:
1. Run `ALTER ROLE postgres SUPERUSER` in the [SQL editor](https://app.supabase.com/project/_/sql).
1. Run `pg_dump --clean --if-exists --quote-all-identifiers -h $OLD_DB_URL -U postgres > dump.sql` from your terminal.
1. Run `ALTER ROLE postgres NOSUPERUSER` in the [SQL editor](https://app.supabase.com/project/_/sql).

In your new project:
1. Run `ALTER ROLE postgres SUPERUSER` in the [SQL editor](https://app.supabase.com/project/_/sql).
1. Run `psql -h $NEW_DB_URL -U postgres -f dump.sql` from your terminal.
1. Run `TRUNCATE storage.objects` in the [SQL editor](https://app.supabase.com/project/_/sql).
1. Run `ALTER ROLE postgres NOSUPERUSER` in the [SQL editor](https://app.supabase.com/project/_/sql).
```

## Migrate storage objects

> This script moves storage objects from one project to another. If you have more than 10k objects, we can move the objects for you. Just contact us at [support@supabase.io](mailto:support@supabase.io).

Reviewer: Let's loop in the caveat here:
```suggestion
The new project has the old project's Storage buckets, but the Storage objects need to be migrated manually. Use this script to move storage objects from one project to another. If you have more than 10k objects, we can move the objects for you. Just contact us at [support@supabase.io](mailto:support@supabase.io).
```

    NEW_PROJECT_URL,
    NEW_PROJECT_SERVICE_KEY
  )

  // make sure you update max_rows in postgrest settings if you have a lot of objects
  // or paginate here
  const { data: oldObjects, error } = await oldSupabaseRestClient
    .from('objects')
    .select()
  if (error) {
    console.log('error getting objects from old bucket')
    throw error
  }

  for (const objectData of oldObjects) {
    console.log(`moving ${objectData.id}`)
    try {
      const { data, error: downloadObjectError } =
        await oldSupabaseClient.storage
          .from(objectData.bucket_id)
          .download(objectData.name)
      if (downloadObjectError) {
        throw downloadObjectError
      }

      const { _, error: uploadObjectError } = await newSupabaseClient.storage
        .from(objectData.bucket_id)
        .upload(objectData.name, data, {
          upsert: true,
          contentType: objectData.metadata.mimetype,
          cacheControl: objectData.metadata.cacheControl,
        })
      if (uploadObjectError) {
        throw uploadObjectError
      }
    } catch (err) {
      console.log('error moving ', objectData)
      console.log(err)
    }
  }
})()
```

> ## Caveats
> 
> - The new project will have the old project's Storage buckets, but not the objects. You will need to migrate Storage objects manually.

Reviewer: Move to section above
```suggestion
```
